### PR TITLE
Meet Ickle, Infinispan's new query language

### PR DESCRIPTION
**Pull request by anistor**
Jira: https://issues.jboss.org/browse/ISPN-6348, https://issues.jboss.org/browse/ISPN-6975 ,
    https://issues.jboss.org/browse/ISPN-7258

See also: http://bit.ly/2f0WrIR

### CODE REVIEW

**Comment by anistor** [comment]
Wait, still need to add the grammar files.

**Comment by anistor** [comment]
Do not pull it yet. I'll squash the last two commits asap, and maybe fix whatever fails in CI.

**Comment by anistor** [comment]
I think I'm done with my changes. Let's see what CI shows.

**Comment by anistor** [comment]
Updated again.

**Comment by gustavonalle** [comment]
looking

File: object-filter/src/main/java/org/infinispan/objectfilter/Matcher.java [hunk 1]
```diff
@@ -7,10 +7,10 @@
 import org.infinispan.query.dsl.Query;
 
 /**
- * An object matcher able to test a given object against multiple registered filters specified either as JPA queries or
- * using the query DSL (see {@link org.infinispan.query.dsl}). The matching filters are notified via a callback supplied
- * when registering the filter. The filter will have to specify the fully qualified type name of the matching
- * object/entity because simple names cannot be resolved as it would happen in the case of an {@link
+ * An object matcher able to test a given object against multiple registered filters specified either as Ickle queries
+ * (a JP-QL subset) or using the query DSL (see {@link org.infinispan.query.dsl}). The matching filters are notified via
```
Comment (gustavonalle) [inline object-filter/src/main/java/org/infinispan/objectfilter/Matcher.java]: Minor: s/a JP-QL subset/a JP-QL subset with full text support/

File: object-filter/src/test/java/org/infinispan/objectfilter/impl/ql/test/TestBase.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+      try {
+         Object[] pair;
+         if (lexerOnly) {
+            pair = executeAndCaptureErrOut(() -> lexer.nextToken());
+         } else {
+            pair = executeAndCaptureErrOut(() -> {
+               IckleParser parser = new IckleParser(tokens);
+               IckleParser.statement_return statement = parser.statement();
+//               System.out.println(tokens.getTokens());
```
Comment (gustavonalle) [inline object-filter/src/test/java/org/infinispan/objectfilter/impl/ql/test/TestBase.java]: remove?

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/parser/package-info.java [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/**
+ * A JP-QL dialect (subset) parser.
```
Comment (gustavonalle) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/parser/package-info.java]: "subset with FT support" would make it clearer?

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/package-info.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /**
- * The syntactic representation of a JPA query. This is the output of the HQL parser. This representation needs further
+ * The syntactic representation of an Infinispan-QL query (the output of the parser). This representation needs further
```
Comment (gustavonalle) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/package-info.java]: s/Infinispan-QL/Ickle/

**Comment by gustavonalle** [comment]
Great stuff @anistor! Just some very minor javadoc and unused stuff.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/parser/package-info.java (hunk 3 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/parser/package-info.java]: yup.

File: object-filter/src/test/java/org/infinispan/objectfilter/impl/ql/test/TestBase.java (hunk 2 above)
Comment (anistor) [inline object-filter/src/test/java/org/infinispan/objectfilter/impl/ql/test/TestBase.java]: yes.

File: object-filter/src/main/java/org/infinispan/objectfilter/Matcher.java (hunk 1 above)
Comment (anistor) [inline object-filter/src/main/java/org/infinispan/objectfilter/Matcher.java]: indeed.

**Comment by gustavonalle** [comment]
@anistor do you need any rebase here, given the recents PRs integrated?

**Comment by anistor** [comment]
yes, need to rebase it and push the fixes. I'm on it.

**Comment by anistor** [comment]
Updated.

**Comment by gustavonalle** [comment]
Integrated, thanks @anistor ! :clap:

**Pull request merged**